File: app/src/main/java/com/sjn/stamp/ui/fragment/media_list/MyStampListFragment.java
package com.sjn.stamp.ui.fragment.media_list;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.content.DialogInterface;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;
import com.sjn.stamp.R;
import com.sjn.stamp.controller.StampController;
import com.sjn.stamp.ui.DialogFacade;
import com.sjn.stamp.ui.SongAdapter;
import com.sjn.stamp.ui.item.SongItem;
import com.sjn.stamp.utils.LogHelper;
import com.sjn.stamp.utils.MediaIDHelper;

import java.util.ArrayList;
import java.util.List;

import eu.davidea.fastscroller.FastScroller;
import eu.davidea.flexibleadapter.FlexibleAdapter;
import eu.davidea.flexibleadapter.common.SmoothScrollLinearLayoutManager;
import eu.davidea.flexibleadapter.helpers.ItemTouchHelperCallback;
import eu.davidea.flexibleadapter.helpers.UndoHelper;
import eu.davidea.flexibleadapter.items.AbstractFlexibleItem;
import eu.davidea.flexibleadapter.items.IFlexible;
import eu.davidea.viewholders.FlexibleViewHolder;

public class MyStampListFragment extends SongListFragment implements
        UndoHelper.OnUndoListener, FlexibleAdapter.OnItemSwipeListener {
    private static final String TAG = LogHelper.INSTANCE.makeLogTag(MyStampListFragment.class);

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        LogHelper.INSTANCE.d(TAG, "onCreateView START" + getMediaId());
        final View rootView = inflater.inflate(R.layout.fragment_list, container, false);

        mLoading = rootView.findViewById(R.id.progressBar);
        mEmptyView = rootView.findViewById(R.id.empty_view);
        mFastScroller = rootView.findViewById(R.id.fast_scroller);
        mEmptyTextView = rootView.findViewById(R.id.empty_text);

        mSwipeRefreshLayout = rootView.findViewById(R.id.refresh);
        mSwipeRefreshLayout.setOnRefreshListener(this);
        mSwipeRefreshLayout.setColorSchemeColors(Color.RED, Color.GREEN, Color.BLUE, Color.YELLOW);

        mAdapter = new SongAdapter(mItemList, this);
        mAdapter.setNotifyChangeOfUnfilteredItems(true)
                .setAnimationOnScrolling(false);
        mRecyclerView = rootView.findViewById(R.id.recycler_view);
        if (savedInstanceState != null) {
            mListState = savedInstanceState.getParcelable(LIST_STATE_KEY);
        }
        RecyclerView.LayoutManager layoutManager = new SmoothScrollLinearLayoutManager(getActivity());
        if (mListState != null) {
            layoutManager.onRestoreInstanceState(mListState);
        }
        mRecyclerView.setLayoutManager(layoutManager);
        mRecyclerView.setAdapter(mAdapter);
        mAdapter.setFastScroller((FastScroller) rootView.findViewById(R.id.fast_scroller));
        mAdapter.setLongPressDragEnabled(false)
                .setHandleDragEnabled(false)
                .setSwipeEnabled(true)
                .setUnlinkAllItemsOnRemoveHeaders(false)
                .setDisplayHeadersAtStartUp(false)
                .setStickyHeaders(false)
                .showAllHeaders();
        initializeFabWithStamp();
        if (mIsVisibleToUser) {
            notifyFragmentChange();
        }
        draw();
        LogHelper.INSTANCE.d(TAG, "onCreateView END");
        return rootView;
    }


    @Override
    public void onItemSwipe(final int position, int direction) {
        LogHelper.INSTANCE.i(TAG, "onItemSwipe position=" + position +
                " direction=" + (direction == ItemTouchHelper.LEFT ? "LEFT" : "RIGHT"));
        final List<Integer> positions = new ArrayList<>(1);
        positions.add(position);
        IFlexible abstractItem = mAdapter.getItem(position);
        final StringBuilder message = new StringBuilder();
        if (abstractItem.isSelectable()) {
            mAdapter.setRestoreSelectionOnUndo(false);
        }
        if (direction == ItemTouchHelper.RIGHT) {
            SongItem subItem = (SongItem) abstractItem;
            StampController stampController = new StampController(getActivity());
            if (stampController.isCategoryStamp(MediaIDHelper.INSTANCE.extractBrowseCategoryValueFromMediaID(mMediaId), false, subItem.getMediaId())) {
                Toast.makeText(getActivity(), R.string.error_message_stamp_failed, Toast.LENGTH_LONG).show();
                final RecyclerView.ViewHolder holder = mRecyclerView.findViewHolderForLayoutPosition(position);
                final View view = ((ItemTouchHelperCallback.ViewHolderCallback) holder).getFrontView();
                Animator animator = ObjectAnimator.ofFloat(view, "translationX", view.getTranslationX(), 0);
                animator.addListener(new Animator.AnimatorListener() {
                    @Override
                    public void onAnimationStart(Animator animator) {

                    }

                    @Override
                    public void onAnimationEnd(Animator animator) {

                    }

                    @Override
                    public void onAnimationCancel(Animator animation) {
                        view.setTranslationX(0);
                    }

                    @Override
                    public void onAnimationRepeat(Animator animator) {

                    }
                });
                animator.start();
                if (holder instanceof FlexibleViewHolder) {
                    ((FlexibleViewHolder) holder).onActionStateChanged(position, ItemTouchHelper.ACTION_STATE_IDLE);
                }
                return;
            }

            message.append(subItem.getTitle()).append(" ");
            DialogFacade.INSTANCE.createRemoveStampSongDialog(getActivity(), subItem.getTitle(), MediaIDHelper.INSTANCE.extractBrowseCategoryValueFromMediaID(mMediaId), new MaterialDialog.SingleButtonCallback() {
                        @Override
                        public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                            switch (which) {
                                case NEGATIVE:
                                    final RecyclerView.ViewHolder holder = mRecyclerView.findViewHolderForLayoutPosition(position);
                                    if (holder instanceof ItemTouchHelperCallback.ViewHolderCallback) {
                                        final View view = ((ItemTouchHelperCallback.ViewHolderCallback) holder).getFrontView();
                                        Animator animator = ObjectAnimator.ofFloat(view, "translationX", view.getTranslationX(), 0);
                                        animator.addListener(new Animator.AnimatorListener() {
                                            @Override
                                            public void onAnimationStart(Animator animator) {

                                            }

                                            @Override
                                            public void onAnimationEnd(Animator animator) {

                                            }

                                            @Override
                                            public void onAnimationCancel(Animator animation) {
                                                view.setTranslationX(0);
                                            }

                                            @Override
                                            public void onAnimationRepeat(Animator animator) {

                                            }
                                        });
                                        animator.start();
                                        if (holder instanceof FlexibleViewHolder) {
                                            ((FlexibleViewHolder) holder).onActionStateChanged(position, ItemTouchHelper.ACTION_STATE_IDLE);
                                        }
                                    }
                                    return;
                                case POSITIVE:
                                    message.append(getString(R.string.action_deleted));
                                    mSwipeRefreshLayout.setRefreshing(true);
                                    new UndoHelper(mAdapter, MyStampListFragment.this)
                                            .withPayload(null) //You can pass any custom object (in this case Boolean is enough)
                                            .withAction(UndoHelper.ACTION_REMOVE, new UndoHelper.SimpleActionListener() {
                                                @Override
                                                public void onPostAction() {
                                                    // Handle ActionMode title
                                                    if (mAdapter.getSelectedItemCount() == 0) {
                                                        mListener.destroyActionModeIfCan();
                                                    } else {
                                                        mListener.updateContextTitle(mAdapter.getSelectedItemCount());
                                                    }
                                                }
                                            })
                                            .remove(positions, getActivity().findViewById(R.id.main_view), message,
                                                    getString(R.string.undo), UndoHelper.UNDO_TIMEOUT);
                                    break;
                            }
                        }
                    },
                    new DialogInterface.OnDismissListener() {
                        @Override
                        public void onDismiss(DialogInterface dialog) {
                            final RecyclerView.ViewHolder holder = mRecyclerView.findViewHolderForLayoutPosition(position);
                            if (holder instanceof ItemTouchHelperCallback.ViewHolderCallback) {
                                final View view = ((ItemTouchHelperCallback.ViewHolderCallback) holder).getFrontView();
                                Animator animator = ObjectAnimator.ofFloat(view, "translationX", view.getTranslationX(), 0);
                                animator.addListener(new Animator.AnimatorListener() {
                                    @Override
                                    public void onAnimationStart(Animator animator) {

                                    }

                                    @Override
                                    public void onAnimationEnd(Animator animator) {

                                    }

                                    @Override
                                    public void onAnimationCancel(Animator animation) {
                                        view.setTranslationX(0);
                                    }

                                    @Override
                                    public void onAnimationRepeat(Animator animator) {

                                    }
                                });
                                animator.start();

                                if (holder instanceof FlexibleViewHolder) {
                                    ((FlexibleViewHolder) holder).onActionStateChanged(position, ItemTouchHelper.ACTION_STATE_IDLE);
                                }
                            }
                        }
                    }).show();
        }
    }

    @Override
    public void onActionStateChanged(RecyclerView.ViewHolder viewHolder, int actionState) {
        LogHelper.INSTANCE.i(TAG, "onActionStateChanged actionState=" + actionState);
        mSwipeRefreshLayout.setEnabled(actionState == ItemTouchHelper.ACTION_STATE_IDLE);
    }

    @Override
    public void onActionCanceled(int action) {
        LogHelper.INSTANCE.i(TAG, "onUndoConfirmed action=" + action);
        if (action == UndoHelper.ACTION_UPDATE) {
        } else if (action == UndoHelper.ACTION_REMOVE) {
            // Custom action is restore deleted items
            mAdapter.restoreDeletedItems();
            // Disable Refreshing
            mSwipeRefreshLayout.setRefreshing(false);
            // Check also selection restoration
            if (mAdapter.isRestoreWithSelection()) {
                mListener.restoreSelection();
            }
        }

    }

    @Override
    public void onActionConfirmed(int action, int event) {
        LogHelper.INSTANCE.i(TAG, "onDeleteConfirmed action=" + action);
        mSwipeRefreshLayout.setRefreshing(false);
        for (AbstractFlexibleItem adapterItem : mAdapter.getDeletedItems()) {
            try {
                switch (adapterItem.getLayoutRes()) {
                    case R.layout.recycler_song_item:
                        SongItem subItem = (SongItem) adapterItem;
                        if (getActivity() != null) {
                            new StampController(getActivity()).removeSong(MediaIDHelper.INSTANCE.extractBrowseCategoryValueFromMediaID(mMediaId), false, subItem.getMediaId());
                        }
                        LogHelper.INSTANCE.i(TAG, "Confirm removed " + subItem.toString());
                        break;
                }

            } catch (IllegalStateException ignored) {
            }
        }
    }

}

File: app/src/main/java/com/sjn/stamp/ui/fragment/media/MyStampListFragment.kt
package com.sjn.stamp.ui.fragment.media

import android.graphics.Color
import android.os.Bundle
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.sjn.stamp.R
import com.sjn.stamp.controller.StampController
import com.sjn.stamp.ui.SongAdapter
import com.sjn.stamp.ui.activity.DrawerActivity
import com.sjn.stamp.ui.item.AbstractItem
import com.sjn.stamp.ui.item.SongItem
import com.sjn.stamp.utils.AlbumArtHelper
import com.sjn.stamp.utils.LogHelper
import com.sjn.stamp.utils.MediaIDHelper
import com.sjn.stamp.utils.SwipeHelper
import eu.davidea.fastscroller.FastScroller
import eu.davidea.flexibleadapter.FlexibleAdapter
import eu.davidea.flexibleadapter.common.SmoothScrollLinearLayoutManager
import eu.davidea.flexibleadapter.helpers.UndoHelper

class MyStampListFragment : SongListFragment(), UndoHelper.OnUndoListener, FlexibleAdapter.OnItemSwipeListener {

    private val categoryValue: String?
        get() = mediaId?.let { MediaIDHelper.extractBrowseCategoryValueFromMediaID(it) }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        LogHelper.d(TAG, "onCreateView START" + mediaId!!)
        val rootView = inflater.inflate(R.layout.fragment_list, container, false)

        if (savedInstanceState != null) {
            listState = savedInstanceState.getParcelable(ListFragment.LIST_STATE_KEY)
        }

        loading = rootView.findViewById(R.id.progressBar)
        emptyView = rootView.findViewById(R.id.empty_view)
        fastScroller = rootView.findViewById(R.id.fast_scroller)
        emptyTextView = rootView.findViewById(R.id.empty_text)
        swipeRefreshLayout = rootView.findViewById(R.id.refresh)
        recyclerView = rootView.findViewById(R.id.recycler_view)

        swipeRefreshLayout?.apply {
            setOnRefreshListener(this@MyStampListFragment)
            setColorSchemeColors(Color.RED, Color.GREEN, Color.BLUE, Color.YELLOW)
        }
        adapter = SongAdapter(currentItems, this).apply {
            setNotifyChangeOfUnfilteredItems(true)
            setAnimationOnScrolling(false)
        }
        recyclerView?.apply {
            activity?.let {
                this.layoutManager = SmoothScrollLinearLayoutManager(it).apply {
                    listState?.let { onRestoreInstanceState(it) }
                }
            }
            this.adapter = this@MyStampListFragment.adapter
        }
        adapter?.apply {
            fastScroller = rootView.findViewById<View>(R.id.fast_scroller) as FastScroller
            isLongPressDragEnabled = false
            isHandleDragEnabled = false
            isSwipeEnabled = true
            setUnlinkAllItemsOnRemoveHeaders(false)
            setDisplayHeadersAtStartUp(false)
            setStickyHeaders(false)
            showAllHeaders()
        }
        initializeFabWithStamp()
        if (isShowing) {
            notifyFragmentChange()
        }
        mediaId?.let {
            MediaIDHelper.extractBrowseCategoryValueFromMediaID(it)?.let {
                arguments?.also {
                    if (activity is DrawerActivity) {
                        (activity as DrawerActivity).run {
                            updateAppbar(it.getString("IMAGE_TEXT"), { activity, imageView ->
                                AlbumArtHelper.loadAlbumArt(activity, imageView, it.getParcelable("IMAGE_BITMAP"), it.getString("IMAGE_TYPE"), it.getString("IMAGE_URL"), it.getString("IMAGE_TEXT"))
                            })
                        }

                    }
                }
            }
        }
        draw()
        LogHelper.d(TAG, "onCreateView END")
        return rootView
    }


    override fun onItemSwipe(position: Int, direction: Int) {
        LogHelper.i(TAG, "onItemSwipe position=" + position +
                " direction=" + if (direction == ItemTouchHelper.LEFT) "LEFT" else "RIGHT")
        if (adapter?.getItem(position) !is SongItem) return
        val item = adapter?.getItem(position) as SongItem
        activity?.run {
            if (categoryValue?.let { StampController(this).isCategoryStamp(it, false, item.mediaId) } == true) {
                Toast.makeText(this, R.string.error_message_stamp_failed, Toast.LENGTH_LONG).show()
                SwipeHelper.cancel(recyclerView, position)
                return
            }
            tryRemove(item, position)
        }
    }

    private fun tryRemove(item: AbstractItem<*>, position: Int) {
        val positions = mutableListOf(position)
        val message = StringBuilder().append(item.title).append(" ").append(getString(R.string.action_deleted))
        if (item.isSelectable) adapter?.setRestoreSelectionOnUndo(false)
        adapter?.isPermanentDelete = false
        swipeRefreshLayout?.isRefreshing = true
        activity?.let {
            UndoHelper(adapter, this@MyStampListFragment)
                    .withPayload(null)
                    .withConsecutive(true)
                    .start(positions, it.findViewById(R.id.main_view), message, getString(R.string.undo), UndoHelper.UNDO_TIMEOUT)
        }
    }


    override fun onActionStateChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
        LogHelper.i(TAG, "onActionStateChanged actionState=$actionState")
        swipeRefreshLayout?.isEnabled = actionState == ItemTouchHelper.ACTION_STATE_IDLE
    }

    override fun onActionCanceled(action: Int) {
        LogHelper.i(TAG, "onUndoConfirmed action=$action")
        adapter?.restoreDeletedItems()
        swipeRefreshLayout?.isRefreshing = false
        if (adapter?.isRestoreWithSelection == true) listener?.restoreSelection()
    }

    override fun onActionConfirmed(action: Int, event: Int) {
        LogHelper.i(TAG, "onDeleteConfirmed action=$action")
        swipeRefreshLayout?.isRefreshing = false
        for (adapterItem in adapter?.deletedItems ?: emptyList()) {
            try {
                when (adapterItem.layoutRes) {
                    R.layout.recycler_song_item -> {
                        val subItem = adapterItem as SongItem
                        activity?.let {
                            StampController(it).run {
                                categoryValue?.let { removeSong(it, false, subItem.mediaId) }
                            }
                            LogHelper.i(TAG, "Confirm removed " + subItem.toString())
                        }
                    }
                }
            } catch (e: IllegalStateException) {
                e.printStackTrace()
            }
        }
    }

    companion object {
        private val TAG = LogHelper.makeLogTag(MyStampListFragment::class.java)
    }

}
